### PR TITLE
Seankeating/intial text component

### PR DESCRIPTION
**Pull request by seankeating**
Added initial Text component
with sizes:
- xs, sm, md, lg, xl, x2, x3, x4

and variations:
- atMention
- disabled
- timestamp

and states:
- error
- success

Additionally added a number of site variables with colors from the Teams color pallet.

### CODE REVIEW

File: docs/src/examples/components/Text/Types/TextCaptionTypeExample.tsx [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import React from 'react'
+import { Text } from 'stardust'
+
+const TextCaptionTypeExample = () => (
+  <Text type="caption">
+    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
```
Comment (dvdzkwsk) [inline docs/src/examples/components/Text/Types/TextCaptionTypeExample.tsx]: We have `faker` available for generating fake text, if you'd like :).

File: src/components/Text/Text.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  error: PropTypes.bool,
+
+  /** Set as success Text component */
+  success: PropTypes.bool,
+
+  /** Set as timestamp Text component */
+  timestamp: PropTypes.bool,
+
+  /** Type of Text component */
+  type: PropTypes.oneOf(['title', 'title2', 'base', 'caption', 'x-small']),
```
Comment (dvdzkwsk) [inline src/components/Text/Text.tsx]: `x-small` but no `small`? I am not privy to these choices.

File: src/components/Text/Text.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import cx from 'classnames'
+import PropTypes from 'prop-types'
+import React from 'react'
+
+import textRules from './textRules'
+
+import { createComponent, customPropTypes, getElementType, getUnhandledProps } from '../../lib'
+
+/**
+ * An component containing text
```
Comment (dvdzkwsk) [inline src/components/Text/Text.tsx]: `An` -> `A`

File: src/components/Text/Text.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+const Text: any = (props: any) => {
+  const ElementType = getElementType(Text, props)
+  const rest = getUnhandledProps(Text, props)
+  const { className, styles } = props
+
+  return <ElementType {...rest} className={cx('ui-text', styles.root, className)} />
+}
+
+Text.defaultProps = {
+  as: 'span',
```
Comment (dvdzkwsk) [inline src/components/Text/Text.tsx:21]: Would `p` possibly be more natural?

File: docs/src/examples/components/Text/Variations/TextExampleAtMention.tsx [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import React from 'react'
+import { Text } from 'stardust'
+
+const TextExampleAtMention = () => <Text atMention>@Russell Wilson</Text>
```
Comment (csano) [inline docs/src/examples/components/Text/Variations/TextExampleAtMention.tsx:4]: Nit: Aaron Judge

File: src/components/Text/Text.tsx (hunk 2 above)
Comment (layershifter) [reply src/components/Text/Text.tsx]: `title` and `title2` look confusing, too

Comment (seankeating) [reply src/components/Text/Text.tsx]: These our from the design document given, I didn't pick these names.

Comment (seankeating) [reply src/components/Text/Text.tsx]: @levithomason - Our those our official names?

File: src/components/Text/textRules.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  title: { fontSize: '24pt', lineHeight: '32pt' },
+  title2: { fontSize: '18pt', lineHeight: '24pt' },
+  base: { fontSize: '14pt', lineHeight: '20pt' },
+  caption: { fontSize: '12pt', lineHeight: '16pt' },
+  'x-small': { fontSize: '10pt', lineHeight: '12pt' },
+}
+
+export default ({ atMention, disabled, error, success, timestamp, type }) => ({
+  root: {
+    ...(atMention && { color: atMentionTextColor }),
```
Comment (dvdzkwsk) [inline src/components/Text/textRules.ts]: It seems important that the rendered element semantically reflect the chosen `type`. For example, rendering some variant of `title` should render some sort of heading tag. Not sure what you all have discussed around this.

File: src/components/Text/Text.tsx (hunk 2 above)
Comment (levithomason) [reply src/components/Text/Text.tsx]: We can and will be updating the names from Teams' design docs to fit the props API.  The site variables and component variables are in question, but we will preserve sane and friendly public APIs.

**EDIT** in other words, no title2's and such :)

File: src/components/Text/Text.tsx (hunk 4 above)
Comment (levithomason) [reply src/components/Text/Text.tsx:21]: A p tag will include default styling that we don't want in many cases.  The Text component is more about styling a bit of text, which is likely to mostly not be a paragraph of text.

Think of these types of cases:

image

Comment (dvdzkwsk) [reply src/components/Text/Text.tsx:21]: Hm. So how will users do paragraphs?

**Comment by levithomason** [comment]
@seankeating do you have any time to wrap this one up?

File: src/components/Text/Text.tsx (hunk 4 above)
Comment (levithomason) [reply src/components/Text/Text.tsx:21]: The user agent styling will handle basic paragraphs:
```jsx
<p>
User paragraph
</p>
```

If you need a paragraph with the special handling of Text:
```jsx
<Text as='p' size='xl' disabled>Do stuff</Text>
```

File: docs/src/examples/components/Text/Types/TextSmallExample.tsx [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import faker from 'faker'
+import React from 'react'
+import { Text } from 'stardust'
+
+const TextSmallExample = () => <Text size="sm">{faker.lorem.paragraph()}</Text>
```
Comment (levithomason) [inline docs/src/examples/components/Text/Types/TextSmallExample.tsx]: Let's merge all the size variations into a single size example.  It should also be moved to "Variations".

File: docs/src/examples/components/Text/Variations/TextExampleAtMention.tsx [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import faker from 'faker'
+import React from 'react'
+import { Text } from 'stardust'
+
+const TextExampleAtMention = () => (
+  <Text atMention>
+    @{faker.name.firstName()} {faker.name.lastName()}
```
Comment (levithomason) [inline docs/src/examples/components/Text/Variations/TextExampleAtMention.tsx]: Just thinking through this, I know we proposed faker for random data, but we will actually need static values as we are planning on doing visual testing against the examples.  You can use faker to get random data, but then bake it in hard coded.

File: test/specs/components/Text/Text-tests.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import { isConformant } from '../../commonTests'
+
+import Text from '../../../../src/components/Text/Text'
```
Comment (levithomason) [inline test/specs/components/Text/Text-tests.ts]: tip: we have root require in tests for ease of dealing with paths.  You can also use the index from your component:

```jsx
import Text from 'src/components/Text'
```

**Review by levithomason — APPROVED** [review]
Looks great, address comments and merge at will! 👍

**Pull request merged**